### PR TITLE
PUBDEV-7274 add MOJO/POJO support for all encodings in GBM

### PR DESCRIPTION
**Pull request by koniecsveta**
https://0xdata.atlassian.net/browse/PUBDEV-7274

added POJO support for binary encoding, eigen encoding and label encoding

added MOJO support for all (OneHotExplicit, SortByResponse, EnumLimited, Enum, Binary, LabelEncoder, Eigen)

### CODE REVIEW

File: h2o-genmodel/src/main/java/hex/genmodel/utils/MathUtils.java [hunk 1]
```diff
@@ -11,4 +11,13 @@ public static boolean equalsWithinRecSumErr(double actual, double expected, int
     return Math.abs(actual - expected) <= (n-1) * Math.ulp(actual) * absum;
   }
 
+  /** Fast calculation of log base 2 for integers.
+   *  @return log base 2 of n */
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/utils/MathUtils.java:15]: incorrectly formatted javadoc

File: h2o-genmodel/src/main/java/hex/genmodel/easy/BinaryColumnMapper.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  public Map<String, Integer> create() {
+    String[] origNames = _m.getOrigNames();
+    String[][] origDomainValues = _m.getOrigDomainValues();
+    Map<String, Integer> columnMapping = new HashMap<>(origNames.length);
+    int pos = 0;
+    // non-categorical
+    for (int i = 0; i < _m.getOrigNumCols(); i++) {
+      if (origDomainValues[i] != null)
+        continue;
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/easy/BinaryColumnMapper.java:25]: the continue statement is confusing why not just use `if (origDomainValues[i]  == null) {}`

File: h2o-genmodel/src/main/java/hex/genmodel/easy/BinaryDomainMapConstructor.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  @Override
+  public Map<Integer, CategoricalEncoder> create() {
+    Map<Integer, CategoricalEncoder> domainMap = new HashMap<>();
+
+    String[] columnNames = _m.getOrigNames();
+    String[][] domainValues = _m.getOrigDomainValues();
+
+    for (int i = 0; i < _m.getOrigNumCols(); i++) {
+      String[] colDomainValues = domainValues[i];
+      int targetOffsetIndex = _columnNameToIndex.get(columnNames[i]);
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/easy/BinaryDomainMapConstructor.java:23]: can be inside the if block

File: h2o-genmodel/src/main/java/hex/genmodel/utils/MathUtils.java (hunk 1 above)
Comment (koniecsveta) [reply h2o-genmodel/src/main/java/hex/genmodel/utils/MathUtils.java]: fixed

File: h2o-genmodel/src/main/java/hex/genmodel/easy/BinaryColumnMapper.java (hunk 2 above)
Comment (koniecsveta) [reply h2o-genmodel/src/main/java/hex/genmodel/easy/BinaryColumnMapper.java]: fixed

File: h2o-genmodel/src/main/java/hex/genmodel/easy/BinaryDomainMapConstructor.java (hunk 3 above)
Comment (koniecsveta) [reply h2o-genmodel/src/main/java/hex/genmodel/easy/BinaryDomainMapConstructor.java]: fixed

File: h2o-algos/src/main/java/hex/tree/SharedTreeModel.java [hunk 4]
```diff
@@ -658,12 +664,32 @@ public SharedTreeSubgraph getSharedTreeSubgraph(final int tidx, final int cls) {
...
     if (encoding != CategoricalEncoding.AUTO) {
       sb.ip("public hex.genmodel.CategoricalEncoding getCategoricalEncoding() { return hex.genmodel.CategoricalEncoding." + 
               encoding.name() + "; }").nl();
     }
     return sb;
   }
+  
+  String toJavaDoubleArray(double[] array) {
+    if (array == null) {
+      return "new double[0]";
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/tree/SharedTreeModel.java:679]: should null just be null?

File: h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java [hunk 5]
```diff
@@ -783,26 +783,50 @@ public static Vec toEigen(Vec src) {
...
+      train = getInteraction(train._key, train.names(), maxLevels).execImpl(dest).get();
     }
     DataInfo dinfo = new DataInfo(train, null, 0, true /*_use_all_factor_levels*/, DataInfo.TransformType.NONE,
             DataInfo.TransformType.NONE, /* skipMissing */ false, /* imputeMissing */ true,
             /* missingBucket */ false, /* weights */ false, /* offset */ false, /* fold */ false, /* intercept */ false);
     DKV.put(dinfo);
     Gram.GramTask gtsk = new Gram.GramTask(null, dinfo).doAll(dinfo._adaptedFrame);
     // round the numbers to float precision to be more reproducible
-//    double[] rounded = gtsk._gram._diag;
+    // double[] rounded = gtsk._gram._diag;
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java:794]: its ok to remove commented out code

File: h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java [hunk 6]
```diff
@@ -783,26 +783,50 @@ public static Vec toEigen(Vec src) {
...
-    Vec v = new ProjectOntoEigenVector(multiple(rounded, (int) gtsk._nobs, 1)).doAll(1, (byte) 3, train).outputFrame().anyVec();
+    double [] array = multiple(rounded, (int) gtsk._nobs, 1);
+    if (created) {
+      train.remove();
+      DKV.remove(source);
+    }
+    return array;
+  }
+  
+  public static Interaction getInteraction(Key<Frame> key, String[] names, int maxLevels) {
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java:807]: maybe a better place for this in Interaction class

Comment (honzasterba) [reply h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java:807]: also this code is duplicated in FrameUtils

File: h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java [hunk 7]
```diff
@@ -783,26 +783,50 @@ public static Vec toEigen(Vec src) {
...
     DKV.put(dinfo);
     Gram.GramTask gtsk = new Gram.GramTask(null, dinfo).doAll(dinfo._adaptedFrame);
     // round the numbers to float precision to be more reproducible
-//    double[] rounded = gtsk._gram._diag;
+    // double[] rounded = gtsk._gram._diag;
     double[] rounded = new double[gtsk._gram._diag.length];
     for (int i = 0; i < rounded.length; ++i)
       rounded[i] = (float) gtsk._gram._diag[i];
     dinfo.remove();
-    Vec v = new ProjectOntoEigenVector(multiple(rounded, (int) gtsk._nobs, 1)).doAll(1, (byte) 3, train).outputFrame().anyVec();
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java:805]: this would make ProjectOntoEigenVector unused and maybe safe to remove

File: h2o-genmodel/src/main/java/hex/genmodel/utils/ParseUtils.java [hunk 8]
```diff
@@ -53,6 +53,7 @@ else if (defVal instanceof Double)
                 return Double.valueOf(input);
         }
 
+        // TODO why do we not expect primitive types as default values since arrays of primitive values are returned?
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/utils/ParseUtils.java:56]: do you plan on fixing this TODO?

File: h2o-genmodel/src/main/java/hex/genmodel/GenModel.java [hunk 9]
```diff
@@ -190,6 +190,10 @@ public int getOrigNumCols() {
     return null;
   }
 
+  /** Returns original Eigen encoder projections array for all columns. */
+  @Override
+  public double[] getOrigProjectionArray() {return new double[]{0};}
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/GenModel.java:195]: why is this here? I do not see it implemented/overriden anywhere

Comment (koniecsveta) [reply h2o-genmodel/src/main/java/hex/genmodel/GenModel.java:195]: it is used in SharedTreeModel.java#toJavaInit() line 666, to pass array which defines the eigen encoder to GenModel

File: h2o-genmodel/src/main/java/hex/genmodel/utils/ParseUtils.java (hunk 8 above)
Comment (koniecsveta) [reply h2o-genmodel/src/main/java/hex/genmodel/utils/ParseUtils.java:56]: no, at least not in this PR. Didn't know what to do with this so put a comment to take o look here. Maybe we can create a Jira ticket with low priority? Or at least leave a comment for somebody who will try to use it to avoid confusion?

File: h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java (hunk 7 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java:805]: it is still used in LinearAlgebraUtils.java#toEigen(), which does similar thing but retruns  Vec v = new ProjectOntoEigenVector( ... ) instead of just double [] array =  multiple( .. ). Array defines encoding and ProjectOntoEigenVector actually encodes the vector by using it.

File: h2o-algos/src/main/java/hex/tree/SharedTreeModel.java (hunk 4 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/tree/SharedTreeModel.java:679]: yes can be 😄 thx

File: h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java (hunk 6 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java:807]: didn't notice, will move, thx

File: h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java (hunk 5 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java]: removed

File: h2o-genmodel/src/main/java/hex/genmodel/utils/ParseUtils.java (hunk 8 above)
Comment (koniecsveta) [reply h2o-genmodel/src/main/java/hex/genmodel/utils/ParseUtils.java]: TODO removed and FIXME comment moved to ParseUtils#tryParse() without FIXME so that it just informs how tryParse() works

File: h2o-genmodel/src/main/java/hex/genmodel/GenModel.java (hunk 9 above)
Comment (koniecsveta) [reply h2o-genmodel/src/main/java/hex/genmodel/GenModel.java]: 👌 changed to return null

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EigenEncoderDomainMapConstructor.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        double[] targetProjectionArray = new double[domainValues.length];
+        System.arraycopy(_m.getOrigProjectionArray() , pos, targetProjectionArray, 0, domainValues.length);
+        pos += domainValues.length;
+        domainMap.put(colIndex, new EigenEncoder(colName, colIndex, domainValues, targetProjectionArray));
+      }
+    }
+    return domainMap;
+  }
+
+  public static long sum(long[] array) {
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/easy/EigenEncoderDomainMapConstructor.java:33]: use ArrayUtils.sum

File: h2o-genmodel/src/main/java/hex/genmodel/easy/LabelEncoderColumnMapper.java [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package hex.genmodel.easy;
+
+import hex.genmodel.GenModel;
+
+
+public class LabelEncoderColumnMapper extends EnumEncoderColumnMapper {
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/easy/LabelEncoderColumnMapper.java:6]: why do we need a special class? can you use just EnumEncoderColumnMapper?

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EigenEncoderDomainMapConstructor.java (hunk 10 above)
Comment (koniecsveta) [reply h2o-genmodel/src/main/java/hex/genmodel/easy/EigenEncoderDomainMapConstructor.java:33]: hm this is not needed at all, it is just some leftover, thx, will remove

File: h2o-genmodel/src/main/java/hex/genmodel/easy/LabelEncoderColumnMapper.java (hunk 11 above)
Comment (koniecsveta) [reply h2o-genmodel/src/main/java/hex/genmodel/easy/LabelEncoderColumnMapper.java:6]: yes I will

File: h2o-algos/src/main/java/hex/tree/SharedTreeMojoWriter.java [hunk 12]
```diff
@@ -38,6 +38,28 @@ protected void writeModelData() throws IOException {
       writekv("calib_method", "platt");
       writekv("calib_glm_beta", beta);
     }
+    writekv("_genmodel_encoding", model.getGenModelEncoding());
+    String[] origNames = model._output._origNames;
+    if (origNames != null) {
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/tree/SharedTreeMojoWriter.java:43]: loot at writeDomains() in AbstractMojoWriter, I think we should handle this in a simmilar manner, same for reading

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java [hunk 13]
```diff
@@ -57,6 +58,14 @@
      * GLM's beta used for calibrating output probabilities using Platt Scaling.
      */
     protected double[] _calib_glm_beta;
+    
+    protected String _genmodel_encoding;
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java:62]: I think these changes are not backwards compatible so you will need to bump the version of all affected MOJOs and update MOJOLand

File: h2o-algos/src/main/java/hex/tree/SharedTreeMojoWriter.java (hunk 12 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/tree/SharedTreeMojoWriter.java:43]: fixed

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java [hunk 14]
```diff
@@ -22,7 +22,38 @@ protected void readModelData() throws IOException {
     _model._ntrees_per_group = tpc;
     _model._compressed_trees = new byte[_model._ntree_groups * tpc][];
     _model._mojo_version = ((Number) readkv("mojo_version")).doubleValue();
-
+    _model._genmodel_encoding = readkv("_genmodel_encoding").toString();
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java:25]: this will throw npe for <1.40 I think

File: h2o-algos/src/main/java/hex/tree/SharedTreeMojoWriter.java [hunk 15]
```diff
@@ -38,6 +38,34 @@ protected void writeModelData() throws IOException {
...
+      }
+      finishWritingTextFile();
+    }
+    if (model._output._origDomains != null) {
+      int nOrigDomainValues = model._output._origDomains.length;
+      writekv("_n_orig_domain_values", nOrigDomainValues);
+      for (int i=0; i < nOrigDomainValues; i++) {
+        String[] currOrigDomain = model._output._origDomains[i];
+        writekv("_m_orig_domain_values_" + i, currOrigDomain == null ? 0 : currOrigDomain.length);
+        if (currOrigDomain != null) {
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/tree/SharedTreeMojoWriter.java:58]: it would be nice to refactor this start - for() - stop into a method, maybe writeDomains() can use it too

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java [hunk 16]
```diff
@@ -22,7 +22,38 @@ protected void readModelData() throws IOException {
...
+        }
+        n = readkv("_n_orig_domain_values");
+        if (n != null) {
+            _model._orig_domain_values = new String[n][];
+            for (int i = 0; i < n; i++) {
+                int m = readkv("_m_orig_domain_values_" + i);
+                if (m > 0) {
+                    String[] orig_domain_values_row = new String[m];
+                    int j = 0;
+                    for (String line : readtext("_orig_domain_values_" + i)) {
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java:47]: should be readText("", true)

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java [hunk 17]
```diff
@@ -22,7 +22,38 @@ protected void readModelData() throws IOException {
...
+                _model._orig_names[i++] = line;
+            }
+        }
+        n = readkv("_n_orig_domain_values");
+        if (n != null) {
+            _model._orig_domain_values = new String[n][];
+            for (int i = 0; i < n; i++) {
+                int m = readkv("_m_orig_domain_values_" + i);
+                if (m > 0) {
+                    String[] orig_domain_values_row = new String[m];
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java:45]: I would also refactor reading of the String array into a reusable method

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java [hunk 18]
```diff
@@ -22,7 +22,37 @@ protected void readModelData() throws IOException {
...
     _model._compressed_trees = new byte[_model._ntree_groups * tpc][];
     _model._mojo_version = ((Number) readkv("mojo_version")).doubleValue();
-
+    if (_model._mojo_version < 1.40) {
+        _model._genmodel_encoding = "AUTO";
+    } else {
+        _model._genmodel_encoding = readkv("_genmodel_encoding").toString();
+        _model._orig_projection_array = readkv("_orig_projection_array");
+        Integer n = readkv("_n_orig_names");
+        if (n != null) {
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java:31]: this is always true, maybe test > 0 ?

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java [hunk 19]
```diff
@@ -22,7 +22,37 @@ protected void readModelData() throws IOException {
...
+            _model._orig_names = new String[n];
+            int i = 0;
+            for (String line : readtext("_orig_names")) {
+                _model._orig_names[i++] = line;
+            }
+        }
+        n = readkv("_n_orig_domain_values");
+        if (n != null) {
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java:39]: same as above

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java [hunk 20]
```diff
@@ -22,7 +22,37 @@ protected void readModelData() throws IOException {
...
+                int m = readkv("_m_orig_domain_values_" + i);
+                if (m > 0) {
+                    String[] orig_domain_values_row = new String[m];
+                    int j = 0;
+                    for (String line : readtext("_orig_domain_values_" + i)) {
+                        orig_domain_values_row[j++] = line;
+                    }
+                    _model._orig_domain_values[i] = orig_domain_values_row;
+                } else {
+                    _model._orig_domain_values[i] = null;
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java:51]: unnecessary

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java (hunk 18 above)
Comment (koniecsveta) [reply h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java:31]: it's not, it is needed like this. For example EnumEncoder doesn't need originalNames, so doesn't have them set, so it writes this as null

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java (hunk 16 above)
Comment (koniecsveta) [reply h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java]: fixed

File: h2o-algos/src/main/java/hex/tree/SharedTreeMojoWriter.java (hunk 15 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/tree/SharedTreeMojoWriter.java:54]: fixed

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java (hunk 17 above)
Comment (koniecsveta) [reply h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java]: fixed

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java (hunk 20 above)
Comment (koniecsveta) [reply h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoReader.java]: fixed

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java (hunk 13 above)
Comment (michalkurka) [reply h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java:62]: Good point, we need to prevent users from running with an old runtime with a newer mojo.
